### ISSUE TITLE
User transaction is not supported with a TransactionSuppressed migrations or a retrying execution strategy Ef core 9

### ISSUE THREAD

**Issue by mortezassaeed** [dotnet/efcore#35263]
I am encountering an issue while upgrading from EF Core 8 to EF Core 9 due to new changes. In my application, which handles highly sensitive data, I need to verify the data after applying migrations.

To achieve this, I need to create a transaction and share it between the DbContext object that runs the migration and the component that checks data integrity. If the data is corrupted, I need to roll back the transaction and undo the migration changes. This code worked seamlessly in EF Core 8, but I am facing an error in EF Core 9.

```C#
foreach (var dbContextMigrationItem in allPendingMigrations)
{
    await dbContextMigrationItem.Key.Database.BeginTransactionAsync();

    foreach (var migartion in dbContextMigrationItem.Value)
    {
        var runner = migrationFactory.Create(migartion, dbContextMigrationItem.Key, logger);
        try
        {
            if (runner == null)
            {
                throw new RunnerNotFoundException($"{migartion} - migration has not runner");
            }

            await runner.Run();
            logger.LogInformation($"{runner.GetType().Name} with {runner.TargetMigration} applied");
        }
        catch (RunnerNotFoundException ex)
        {
            logger.LogError(ex.Message);
            throw;
        }
        catch (Exception ex)
        {
            logger.LogError($"{runner.GetType().Name} with {runner.TargetMigration} failed");
            logger.LogError(ex.Message);
            throw;
        }
    }

    await dbContextMigrationItem.Key.Database.CommitTransactionAsync();
}
```

The code below is a sample of a migration runner and the MigrationTools class.

```C#
public abstract class MigrationRunnerBase
{
    public DbContext DbContext { get; }
    public string TargetMigration => GetType().GetCustomAttribute<TargetMigrationAttribute>()!.TargetMigrationName;
    protected SqlMigrationTool MigrationTool { get; }
    protected ILogger Logger;

    public MigrationRunnerBase(DbContext dbContext, ILogger logger)
    {
        DbContext = dbContext;
        MigrationTool = SqlMigrationToolFactory.Create(dbContext);
        Logger = logger;
    }

    public abstract Task Run();
}
```
```C#
[TargetMigration("20241203105436_add-sample-entity")]
public class Migration_20241203105436 : MigrationRunnerBase
{
    public Migration_20241203105436(DbContext dbContext, ILogger logger) : base(dbContext, logger)
    {

    }

    public override async Task Run()
    {
        await MigrationTool.RunMigration(TargetMigration);

        // Here, check the data for correctness and throw an exception if something is wrong with the data
    }
}
```
```C#
public class SqlMigrationTool
{
    private readonly DbContext _dbContext;
    private readonly IMigrator _migrator;
    private List<string>? _pendingMigrations;

    public string DbContextFullName { get; }

    public SqlMigrationTool(DbContext dbContext)
    {
        _dbContext = dbContext;
        _migrator = dbContext.Database.GetService<IMigrator>();
        DbContextFullName = dbContext.GetType().FullName!;
    }

    public async Task<bool> IsMigrationAppliedBefore(string targetMigration, DbContext dbContext)
    {
        if (_pendingMigrations == null)
        {
            _pendingMigrations = (await dbContext.Database.GetPendingMigrationsAsync()).ToList();
        }

        return !_pendingMigrations.Any(m => m == targetMigration);
    }

    public async Task RunMigration(string targetMigration)
    {
        if (await IsMigrationAppliedBefore(targetMigration, _dbContext))
        {
            return;
        }

        await _migrator.MigrateAsync(targetMigration);
    }
}
```
So, is it reasonable to stay with .net 8? will there be any possible solutions for this in EF Core 10? are there any alternative approaches?

**Comment by AndriySvyryd** [comment]
Duplicate of https://github.com/dotnet/efcore/issues/35127
Should be fixed in 9.0.1

**Comment by mortezassaeed** [comment]
When can we expect to get version 9.0.1?
